File: src/security.py
"""
Простая и понятная система безопасности для Claude Code.

Принципы:
1. Агент может работать везде (нет workspace boundary)
2. Предупреждаем о системных путях и деструктивных командах
3. Всегда спрашиваем подтверждение перед действием
"""
from __future__ import annotations

import re
from dataclasses import dataclass
from enum import Enum
from pathlib import Path


class ActionRisk(Enum):
    """Уровень риска действия."""
    SAFE = "safe"           # Безопасные действия (чтение, ls)
    LOW = "low"             # Низкий риск (создание файлов)
    MEDIUM = "medium"       # Средний риск (редактирование, перемещение)
    HIGH = "high"           # Высокий риск (удаление, системные пути)
    CRITICAL = "critical"   # Критический риск (rm -rf, форматирование)


@dataclass
class SecurityCheck:
    """Результат проверки безопасности."""
    allowed: bool
    risk: ActionRisk
    warning: str | None = None
    reason: str | None = None


class SecurityValidator:
    """Валидатор безопасности действий."""

    # Системные пути (высокий риск)
    SYSTEM_PATHS = [
        "/etc/", "/sys/", "/proc/", "/dev/", "/boot/",
        "/bin/", "/sbin/", "/usr/bin/", "/usr/sbin/"
    ]

    # Деструктивные команды (критический риск)
    DESTRUCTIVE_COMMANDS = [
        r"\brm\s+-rf\b",
        r"\bdd\b.*if=",
        r"\bmkfs\b",
        r"\bformat\b",
        r"\bfdisk\b",
        r"\bparted\b",
        r"\b:\(\)\{.*\};\s*:\b",  # Fork bomb
    ]

    # Опасные команды (высокий риск)
    DANGEROUS_COMMANDS = [
        r"\brm\b",
        r"\bmv\b.*\s+/",
        r"\bchmod\b.*777",
        r"\bchown\b.*root",
        r"\bsudo\b",
        r"\bsu\b",
    ]

    # Катастрофические команды — БЕЗУСЛОВНЫЙ блок, без вариантов.
    # В отличие от DESTRUCTIVE_COMMANDS выше (которые только предупреждают,
    # allowed=True), эти паттерны нарочно узкие и целятся именно в
    # необратимое уничтожение системы/диска целиком — а не в обычный
    # "rm -rf ./build", который агенту нужен постоянно и который блокировать
    # нельзя. Это последний рубеж на случай опечатки, галлюцинации модели
    # или неправильно распознанной голосовой команды — срабатывает всегда,
    # независимо от auto_approve и от того, есть ли рядом живой человек за
    # терминалом, чтобы нажать 'y'.
    HARD_BLOCK_PATTERNS = [
        (r"\brm\s+.*-[a-z]*r[a-z]*f[a-z]*\s+/(\s|$)", "rm -rf на корень файловой системы"),
        (r"\brm\s+.*-[a-z]*r[a-z]*f[a-z]*\s+/\*", "rm -rf на всё содержимое корня"),
        (r"\brm\s+.*-[a-z]*r[a-z]*f[a-z]*\s+(~|\$HOME)(\s|/|$)", "rm -rf на домашнюю папку целиком"),
        (r"--no-preserve-root", "обход защиты rm от удаления корня"),
        (r":\(\)\s*\{\s*:\s*\|\s*:\s*&\s*\}\s*;\s*:", "fork bomb"),
        (r"\bdd\b[^\n|]*\bof=/dev/(sd|nvme|hd|vd|xvd)", "прямая запись поверх диска целиком"),
        (r"\bmkfs(\.\w+)?\b", "форматирование раздела/диска"),
        (r"\bwipefs\b", "стирание разметки диска"),
        (r">\s*/dev/(sd|nvme|hd|vd|xvd)[a-z0-9]*(\s|$)", "запись напрямую в блочное устройство"),
        (r"\bchmod\s+-R\s+777\s+/(\s|$)", "рекурсивный chmod 777 на корень"),
    ]

    def check_catastrophic(self, command: str) -> str | None:
        """Проверить, не является ли команда безусловно недопустимой.

        Возвращает причину блокировки или None, если команда не входит
        в список необратимых системных катастроф (сама по себе НЕ
        заменяет обычную оценку риска ниже — это отдельный, более узкий
        и более жёсткий барьер).
        """
        for pattern, reason in self.HARD_BLOCK_PATTERNS:
            if re.search(pattern, command, re.IGNORECASE):
                return reason
        return None

    def check_bash_command(self, command: str) -> SecurityCheck:
        """
        Проверить bash команду.

        Args:
            command: Команда для проверки

        Returns:
            SecurityCheck с результатом
        """
        # Проверка на деструктивные команды
        for pattern in self.DESTRUCTIVE_COMMANDS:
            if re.search(pattern, command, re.IGNORECASE):
                return SecurityCheck(
                    allowed=True,  # Разрешаем, но с предупреждением
                    risk=ActionRisk.CRITICAL,
                    warning="⚠️  КРИТИЧЕСКАЯ ОПАСНОСТЬ: Деструктивная команда!",
                    reason=f"Команда может уничтожить данные: {command}"
                )

        # Проверка на опасные команды
        for pattern in self.DANGEROUS_COMMANDS:
            if re.search(pattern, command, re.IGNORECASE):
                return SecurityCheck(
                    allowed=True,
                    risk=ActionRisk.HIGH,
                    warning="⚠️  ВЫСОКИЙ РИСК: Опасная команда!",
                    reason=f"Команда может изменить систему: {command}"
                )

        # Безопасные команды для чтения
        safe_commands = ["ls", "cat", "head", "tail", "grep", "find", "pwd", "echo", "date"]
        if any(command.strip().startswith(cmd) for cmd in safe_commands):
            return SecurityCheck(
                allowed=True,
                risk=ActionRisk.SAFE,
            )

        # По умолчанию - низкий риск
        return SecurityCheck(
            allowed=True,
            risk=ActionRisk.LOW,
        )

    def check_file_path(self, file_path: str, action: str = "access") -> SecurityCheck:
        """
        Проверить путь к файлу.

        Args:
            file_path: Путь к файлу
            action: Тип действия (read, write, delete)

        Returns:
            SecurityCheck с результатом
        """
        # Проверка на системные пути
        for sys_path in self.SYSTEM_PATHS:
            if file_path.startswith(sys_path):
                return SecurityCheck(
                    allowed=True,
                    risk=ActionRisk.HIGH,
                    warning=f"⚠️  СИСТЕМНЫЙ ПУТЬ: {sys_path}",
                    reason=f"Изменение системных файлов может сломать систему"
                )

        # Проверка на чувствительные файлы
        sensitive_patterns = [
            r"/\.ssh/",
            r"/\.aws/",
            r"/\.env$",
            r"password",
            r"secret",
            r"token",
            r"credentials",
            r"id_rsa",
            r"id_ed25519",
        ]

        for pattern in sensitive_patterns:
            if re.search(pattern, file_path, re.IGNORECASE):
                return SecurityCheck(
                    allowed=True,
                    risk=ActionRisk.HIGH,
                    warning="⚠️  ЧУВСТВИТЕЛЬНЫЙ ФАЙЛ!",
                    reason=f"Файл может содержать секреты: {file_path}"
                )

        # Определяем риск по типу действия
        if action == "read":
            risk = ActionRisk.SAFE
        elif action == "write":
            risk = ActionRisk.LOW
        elif action == "delete":
            risk = ActionRisk.MEDIUM
        else:
            risk = ActionRisk.LOW

        return SecurityCheck(
            allowed=True,
            risk=risk,
        )

    def check_file_write(self, file_path: str, content: str) -> SecurityCheck:
        """Проверить запись файла."""
        return self.check_file_path(file_path, action="write")

    def check_file_read(self, file_path: str) -> SecurityCheck:
        """Проверить чтение файла."""
        return self.check_file_path(file_path, action="read")

    def check_file_delete(self, file_path: str) -> SecurityCheck:
        """Проверить удаление файла."""
        check = self.check_file_path(file_path, action="delete")

        # Удаление всегда как минимум средний риск
        if check.risk == ActionRisk.SAFE or check.risk == ActionRisk.LOW:
            check.risk = ActionRisk.MEDIUM
            check.warning = "⚠️  Удаление файла"

        return check


# Глобальный экземпляр валидатора
_validator = SecurityValidator()


def check_bash_command(command: str) -> SecurityCheck:
    """Проверить bash команду."""
    return _validator.check_bash_command(command)


def check_file_write(file_path: str, content: str = "") -> SecurityCheck:
    """Проверить запись файла."""
    return _validator.check_file_write(file_path, content)


def check_file_read(file_path: str) -> SecurityCheck:
    """Проверить чтение файла."""
    return _validator.check_file_read(file_path)


def check_file_delete(file_path: str) -> SecurityCheck:
    """Проверить удаление файла."""
    return _validator.check_file_delete(file_path)

File: src/security.rs
use regex::Regex;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ActionRisk {
    Safe,
    Low,
    Medium,
    High,
    Critical,
}

#[derive(Debug, Clone)]
pub struct SecurityCheck {
    pub allowed: bool,
    pub risk: ActionRisk,
    pub warning: Option<String>,
    pub reason: Option<String>,
}

impl SecurityCheck {
    pub fn allowed(risk: ActionRisk) -> Self {
        Self { allowed: true, risk, warning: None, reason: None }
    }

    pub fn with_warning(mut self, warning: &str) -> Self {
        self.warning = Some(warning.to_string());
        self
    }

    pub fn with_reason(mut self, reason: &str) -> Self {
        self.reason = Some(reason.to_string());
        self
    }
}

pub struct SecurityValidator {
    system_paths: Vec<String>,
    destructive_patterns: Vec<Regex>,
    dangerous_patterns: Vec<Regex>,
    hard_block_patterns: Vec<(Regex, String)>,
    safe_commands: Vec<String>,
    sensitive_patterns: Vec<Regex>,
}

impl SecurityValidator {
    pub fn new() -> Self {
        let windows = cfg!(windows);
        let system_paths: Vec<String> = if windows {
            vec![
                "c:\\windows\\".to_string(), "c:\\program files\\".to_string(),
                "c:\\program files (x86)\\".to_string(), "c:\\boot\\".to_string(),
            ]
        } else {
            vec![
                "/etc/".to_string(), "/sys/".to_string(), "/proc/".to_string(),
                "/dev/".to_string(), "/boot/".to_string(), "/bin/".to_string(),
                "/sbin/".to_string(), "/usr/bin/".to_string(), "/usr/sbin/".to_string(),
            ]
        };

        // Деструктивные паттерны: POSIX и Windows наборы объединены —
        // валидатор один, а команды могут прийти в любой оболочке
        // (WSL-пути, вызов rm из git-bash и т.п.).
        let mut destructive_patterns: Vec<Regex> = vec![
            r"\brm\s+-rf\b",
            r"\bdd\b.*if=",
            r"\bmkfs\b",
            r"\bformat\b",
            r"\bfdisk\b",
            r"\bparted\b",
            r":\(\)\{.*\};\s*:",
        ].into_iter().map(|p| Regex::new(p).unwrap()).collect();
        // WINDOWS: PowerShell/cmd деструктив.
        destructive_patterns.extend([
            r"(?i)Remove-Item\s+[^;\n]*-Recurse[^;\n]*-Force",
            r"(?i)rd\s+/s\s+/q",
            r"(?i)del\s+/s\s+/q",
            r"(?i)reg\s+(delete|add)\s+HKLM",
            r"(?i)bcdedit\s+/set",
            r"(?i)vssadmin\s+delete\s+shadows",
            r"(?i)shutdown\s+/[rts]",
        ].into_iter().map(|p| Regex::new(p).unwrap()));

        let mut dangerous_patterns: Vec<Regex> = vec![
            // ИСПРАВЛЕНО (жалоба "часто ERROR Bash даже на обычных
            // командах"): было `\brm\b` и `\bmv\b.*\s+/` — ловили
            // АБСОЛЮТНО любое использование `rm`/`mv`, даже безобидное
            // `rm file.txt` или `mv a.txt b/c.txt` (у mv практически
            // любой путь назначения содержит "/", так что паттерн
            // срабатывал почти всегда). Это не агент ошибался — это
            // код требовал `force: true` на самые обычные операции,
            // которые агент разумно не помечал как опасные. Настоящая
            // опасность `rm` (рекурсивно/принудительно) уже отдельно
            // ловится ниже в destructive_patterns (`rm -rf` → Critical)
            // и в hard_block_patterns (rm -rf на корень/домашнюю папку
            // → безусловный блок) — здесь дублирование только вредило.
            r"\brm\s+.*-[a-z]*r",   // rm с флагом -r/-rf в любом порядке флагов, но не 'rm -rf /' (уже Critical) — Medium-риск для остальных рекурсивных rm
            r"\bmv\b.*\s+(/etc|/sys|/proc|/dev|/boot|/bin|/sbin|/usr/bin|/usr/sbin)(/|\s|$)", // перемещение В системный путь, а не любой путь со слэшем
            r"\bchmod\b.*777",
            r"\bchown\b.*root",
            r"\bsudo\b",
            r"\bsu\b",
        ].into_iter().map(|p| Regex::new(p).unwrap()).collect();
        // WINDOWS: потенциально системные операции.
        dangerous_patterns.extend([
            r"(?i)Set-ExecutionPolicy",
            r"(?i)Remove-Item\s+[^;\n]*(C:\\Windows|C:\\Program Files)",
            r"(?i)net\s+user\s+\S+\s+/delete",
            r"(?i)taskkill\s+/f\s+/im\s+(explorer|csrss|wininit|svchost)",
        ].into_iter().map(|p| Regex::new(p).unwrap()));

        let mut hard_block_patterns: Vec<(Regex, String)> = vec![
            (r"\brm\s+.*-[a-z]*r[a-z]*f[a-z]*\s+/(\s|$)", "rm -rf на корень файловой системы"),
            (r"\brm\s+.*-[a-z]*r[a-z]*f[a-z]*\s+/\*", "rm -rf на всё содержимое корня"),
            (r"\brm\s+.*-[a-z]*r[a-z]*f[a-z]*\s+(~|\$HOME)(\s|/|$)", "rm -rf на домашнюю папку целиком"),
            (r"--no-preserve-root", "обход защиты rm от удаления корня"),
            (r":\(\)\s*\{\s*:\s*\|\s*:\s*&\s*\}\s*;\s*:", "fork bomb"),
            (r"\bdd\b[^\n|]*\bof=/dev/(sd|nvme|hd|vd|xvd)", "прямая запись поверх диска целиком"),
            (r"\bmkfs(\.\w+)?\b", "форматирование раздела/диска"),
            (r"\bwipefs\b", "стирание разметки диска"),
            (r">\s*/dev/(sd|nvme|hd|vd|xvd)[a-z0-9]*(\s|$)", "запись напрямую в блочное устройство"),
            (r"\bchmod\s+-R\s+777\s+/(\s|$)", "рекурсивный chmod 777 на корень"),
        ].into_iter().map(|(p, r)| (Regex::new(p).unwrap(), r.to_string())).collect();
        // WINDOWS: безусловные блоки.
        hard_block_patterns.extend([
            (r"(?i)Remove-Item\s+[^;\n]*-Recurse[^;\n]*-Force[^;\n]*\s+C:\\(\s|$)", "Remove-Item -Recurse -Force на корень диска C:\\"),
            (r"(?i)rd\s+/s\s+/q\s+(c:\\|c:$)", "rd /s /q на корень диска"),
            (r"(?i)diskpart", "diskpart — прямая работа с разделами диска"),
            (r"(?i)format\s+[a-z]:\s", "форматирование диска"),
            (r"(?i)cipher\s+/w", "cipher /w — затирание свободного места необратимо"),
        ].into_iter().map(|(p, r)| (Regex::new(p).unwrap(), r.to_string())));

        let safe_commands: Vec<String> = if windows {
            vec![
                "Get-ChildItem", "Get-Content", "Get-Location", "Write-Output", "Get-Date",
                "Select-String", "Get-Process", "dir", "type", "echo", "cls",
                "ls", "cat", "grep", "find", "pwd", "date", "git", "cargo",
            ].into_iter().map(String::from).collect()
        } else {
            vec![
                "ls", "cat", "head", "tail", "grep", "find", "pwd", "echo", "date",
                "git", "cargo",
            ].into_iter().map(String::from).collect()
        };

        let sensitive_patterns: Vec<Regex> = vec![
            r"/\.ssh/", r"/\.aws/", r"/\.env$",
            r"password", r"secret", r"token", r"credentials",
            r"id_rsa", r"id_ed25519",
            // WINDOWS-пути секретов.
            r"(?i)\\.ssh\\", r"(?i)\\.aws\\", r"(?i)\\.env$",
        ].into_iter().map(|p| Regex::new(p).unwrap()).collect();

        let _ = windows;
        Self {
            system_paths,
            destructive_patterns,
            dangerous_patterns,
            hard_block_patterns,
            safe_commands,
            sensitive_patterns,
        }
    }

    pub fn check_catastrophic(&self, command: &str) -> Option<String> {
        for (pattern, reason) in &self.hard_block_patterns {
            if pattern.is_match(command) {
                return Some(reason.clone());
            }
        }
        None
    }

    pub fn check_bash_command(&self, command: &str) -> SecurityCheck {
        if let Some(reason) = self.check_catastrophic(command) {
            return SecurityCheck {
                allowed: false,
                risk: ActionRisk::Critical,
                warning: Some(" БЕЗУСЛОВНАЯ БЛОКИРОВКА!".to_string()),
                reason: Some(reason),
            };
        }

        for pattern in &self.destructive_patterns {
            if pattern.is_match(command) {
                return SecurityCheck::allowed(ActionRisk::Critical)
                    .with_warning("⚠️ КРИТИЧЕСКАЯ ОПАСНОСТЬ: Деструктивная команда!")
                    .with_reason(&format!("Команда может уничтожить данные: {}", command));
            }
        }

        for pattern in &self.dangerous_patterns {
            if pattern.is_match(command) {
                return SecurityCheck::allowed(ActionRisk::High)
                    .with_warning("⚠️ ВЫСОКИЙ РИСК: Опасная команда!")
                    .with_reason(&format!("Команда может изменить систему: {}", command));
            }
        }

        let cmd_name = command.trim().split_whitespace().next().unwrap_or("");
        if self.safe_commands.iter().any(|c| c == cmd_name) {
            return SecurityCheck::allowed(ActionRisk::Safe);
        }

        SecurityCheck::allowed(ActionRisk::Low)
    }

    pub fn check_file_path(&self, file_path: &str, action: &str) -> SecurityCheck {
        for sys_path in &self.system_paths {
            if file_path.starts_with(sys_path) {
                return SecurityCheck::allowed(ActionRisk::High)
                    .with_warning(&format!("⚠️ СИСТЕМНЫЙ ПУТЬ: {}", sys_path))
                    .with_reason("Изменение системных файлов может сломать систему");
            }
        }

        for pattern in &self.sensitive_patterns {
            if pattern.is_match(file_path) {
                return SecurityCheck::allowed(ActionRisk::High)
                    .with_warning("⚠️ ЧУВСТВИТЕЛЬНЫЙ ФАЙЛ!")
                    .with_reason(&format!("Файл может содержать секреты: {}", file_path));
            }
        }

        let risk = match action {
            "read" => ActionRisk::Safe,
            "write" => ActionRisk::Low,
            "delete" => ActionRisk::Medium,
            _ => ActionRisk::Low,
        };

        SecurityCheck::allowed(risk)
    }

    pub fn check_file_write(&self, file_path: &str, _content: &str) -> SecurityCheck {
        self.check_file_path(file_path, "write")
    }

    pub fn check_file_read(&self, file_path: &str) -> SecurityCheck {
        self.check_file_path(file_path, "read")
    }

    pub fn check_file_delete(&self, file_path: &str) -> SecurityCheck {
        let mut check = self.check_file_path(file_path, "delete");
        if check.risk == ActionRisk::Safe || check.risk == ActionRisk::Low {
            check.risk = ActionRisk::Medium;
            check.warning = Some("⚠️ Удаление файла".to_string());
        }
        check
    }
}

impl Default for SecurityValidator {
    fn default() -> Self {
        Self::new()
    }
}